Exit with error if --outfile cannot be written

karkhaz opened the pull request:
This commit prevents CBMC from exiting with 0 if the file specified to
`cbmc --outfile` cannot be written. This allows build systems that
contain commands that depend on the file being written to abort the
build if the file was not written.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

karkhaz committed:
Exit with error if --outfile cannot be written

Move file-opening logic from bv_dimacst to solver_factoryt::get_dimacs()
so that errors are handled at the caller level rather than via exit() in
a destructor. bv_dimacst now takes an std::ostream& instead of a
filename, and the caller uses open_outfile_and_check() which throws
invalid_command_line_argument_exceptiont on failure. Moved the
definition of `open_outfile_and_check` further up the file to enable
re-use.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig reviewed (changes requested):
Thank you for calling out this problem and coming up with the smallest possible patch! I think, however, we should go for a more intrusive change and move the entire file-opening logic to the caller. They will be in a much better place to handle errors, and we could thereby avoid doing an `exit` buried deep in the call tree.

I'd suggest that all the code from `bv_dimacst::write_dimacs()` be moved to the only caller that's in `src/goto-checker/solver_factory.cpp`.

karkhaz commented:
I've added a test and fixed the formatting.

@tautschnig that makes sense, but I could use a little help here. As far as I can see, this code is called from the `bv_dimacst` destructor only, so I can't see how to call it from `solver_factory::get_dimacs()` since that method actually leaks the object. It seems that the destructor is never called until right at the end of the program, if I am interpreting this stack trace correctly (I set a breakpoint in lldb and dumped the stack):
```
* thread #1, queue = 'com.apple.main-thread', stop reason = step in
    frame #0: 0x0000000100da9527 cbmc`bv_dimacst::write_dimacs(this=0x0000000106809800) at bv_dimacs.cpp:28:5
   25
   26  	  if(!out)
   27  	  {
-> 28  	    log.error() << "failed to open " << filename << messaget::eom;
   29  	    return true;
   30  	  }
   31
Target 0: (cbmc) stopped.
(lldb) thread list
Process 29968 stopped
* thread #1: tid = 0x26959, 0x0000000100da9527 cbmc`bv_dimacst::write_dimacs(this=0x0000000106809800) at bv_dimacs.cpp:28:5, queue = 'com.apple.main-thread', stop reason = step in
(lldb) thread backtrace
* thread #1, queue = 'com.apple.main-thread', stop reason = step in
  * frame #0: 0x0000000100da9527 cbmc`bv_dimacst::write_dimacs(this=0x0000000106809800) at bv_dimacs.cpp:28:5
    frame #1: 0x00000001000ebc23 cbmc`bv_dimacst::~bv_dimacst(this=0x0000000106809800) at bv_dimacs.h:31:5
    frame #2: 0x00000001000e3955 cbmc`bv_dimacst::~bv_dimacst(this=0x0000000106809800) at bv_dimacs.h:30:3
    frame #3: 0x00000001000e3979 cbmc`bv_dimacst::~bv_dimacst(this=0x0000000106809800) at bv_dimacs.h:30:3
    frame #4: 0x0000000100035b5c cbmc`std::__1::default_delete<decision_proceduret>::operator(this=0x0000000106405870, __ptr=0x0000000106809800)(decision_proceduret*) const at memory:1428:5
    frame #5: 0x0000000100035adc cbmc`std::__1::unique_ptr<decision_proceduret, std::__1::default_delete<decision_proceduret> >::reset(this=0x0000000106405870, __p=0x0000000000000000) at memory:1689:7
    frame #6: 0x0000000100035a79 cbmc`std::__1::unique_ptr<decision_proceduret, std::__1::default_delete<decision_proceduret> >::~unique_ptr(this=0x0000000106405870) at memory:1643:19
    frame #7: 0x0000000100035a15 cbmc`std::__1::unique_ptr<decision_proceduret, std::__1::default_delete<decision_proceduret> >::~unique_ptr(this=0x0000000106405870) at memory:1643:17
    frame #8: 0x00000001000359dd cbmc`solver_factoryt::solvert::~solvert(this=0x0000000106405860) at solver_factory.h:39:9
    frame #9: 0x00000001000359b5 cbmc`solver_factoryt::solvert::~solvert(this=0x0000000106405860) at solver_factory.h:39:9
    frame #10: 0x000000010003596b cbmc`std::__1::default_delete<solver_factoryt::solvert>::operator(this=0x00007ffeefbfd888, __ptr=0x0000000106405860)(solver_factoryt::solvert*) const at memory:1428:5
    frame #11: 0x00000001000358ec cbmc`std::__1::unique_ptr<solver_factoryt::solvert, std::__1::default_delete<solver_factoryt::solvert> >::reset(this=0x00007ffeefbfd888, __p=0x0000000000000000) at memory:1689:7
    frame #12: 0x0000000100035889 cbmc`std::__1::unique_ptr<solver_factoryt::solvert, std::__1::default_delete<solver_factoryt::solvert> >::~unique_ptr(this=0x00007ffeefbfd888) at memory:1643:19
    frame #13: 0x00000001000350f5 cbmc`std::__1::unique_ptr<solver_factoryt::solvert, std::__1::default_delete<solver_factoryt::solvert> >::~unique_ptr(this=0x00007ffeefbfd888) at memory:1643:17
    frame #14: 0x00000001000350ba cbmc`goto_symex_property_decidert::~goto_symex_property_decidert(this=0x00007ffeefbfd870) at goto_symex_property_decider.h:23:7
    frame #15: 0x0000000100035085 cbmc`goto_symex_property_decidert::~goto_symex_property_decidert(this=0x00007ffeefbfd870) at goto_symex_property_decider.h:23:7
    frame #16: 0x0000000100036890 cbmc`multi_path_symex_checkert::~multi_path_symex_checkert(this=0x00007ffeefbfcf98) at multi_path_symex_checker.h:25:7
    frame #17: 0x0000000100036685 cbmc`multi_path_symex_checkert::~multi_path_symex_checkert(this=0x00007ffeefbfcf98) at multi_path_symex_checker.h:25:7
    frame #18: 0x000000010003665e cbmc`stop_on_fail_verifiert<multi_path_symex_checkert>::~stop_on_fail_verifiert(this=0x00007ffeefbfce30) at stop_on_fail_verifier.h:22:7
    frame #19: 0x0000000100023425 cbmc`stop_on_fail_verifiert<multi_path_symex_checkert>::~stop_on_fail_verifiert(this=0x00007ffeefbfce30) at stop_on_fail_verifier.h:22:7
    frame #20: 0x00000001000214f4 cbmc`cbmc_parse_optionst::doit(this=0x00007ffeefbff068) at cbmc_parse_options.cpp:570:5
    frame #21: 0x0000000101402e66 cbmc`parse_options_baset::main(this=0x00007ffeefbff068) at parse_options.cpp:100:12
    frame #22: 0x000000010000437f cbmc`main(argc=5, argv=0x00007ffeefbff320) at cbmc_main.cpp:48:27
    frame #23: 0x00007fff2040af3d libdyld.dylib`start + 1
```
It's entirely possible that my C++ is rusty after all this time but it seems to me that this `bv_dimacst` object is referenced by a unique pointer, which persists until basically the program terminates?

tautschnig commented:
Thank you for digging up the details of the call stack! I think doing the write in the destructor is a design flaw, not a feature. We should seek to use the same approach as what `--smt2 --outfile bla` ends up doing. In that case, the writing is taken care of by the `dec_solve` method (and the file is opened in `solver_factoryt::get_smt2`).

martin-cs commented on regression/cbmc/outfile_no_dir/test.desc:
```diff
@@ -0,0 +1,7 @@
+CORE paths-lifo-expected-failure
+main.c
+--dimacs --outfile /xyz/out.dimacs
```
What will happen on Windows?

martin-cs reviewed (approved):
Concept is great.  Fix is great.  PR is great.
BUT...
I think I agree with @tautschnig about something more radical being needed.

kroening commented:
I realise I am late to this party. Do we want to promote https://github.com/diffblue/hw-cbmc/blob/main/src/ebmc/output_file.h  to cbmc?

tautschnig commented:
> I realise I am late to this party. Do we want to promote https://github.com/diffblue/hw-cbmc/blob/main/src/ebmc/output_file.h to cbmc?

Yes: https://github.com/diffblue/hw-cbmc/pull/588#issuecomment-2221109708 :-)
